internal/pkg/clock/chainclock.go

```diff
@@ -77,18 +86,34 @@ func (cc *chainClock) StartTimeOfEpoch(e abi.ChainEpoch) time.Time {
 	return addedTime
 }
 
-// WaitNextEpoch returns after the next epoch occurs or ctx is done.
+// WaitNextEpoch returns after the next epoch occurs, or ctx is done.
 func (cc *chainClock) WaitNextEpoch(ctx context.Context) abi.ChainEpoch {
 	currEpoch := cc.EpochAtTime(cc.Now())
 	nextEpoch := currEpoch + 1
-	nextEpochStart := cc.StartTimeOfEpoch(nextEpoch)
+	cc.WaitForEpoch(ctx, nextEpoch)
+	return nextEpoch
+}
+
+// WaitForEpoch returns when an epoch is due to start, or ctx is done.
+func (cc *chainClock) WaitForEpoch(ctx context.Context, e abi.ChainEpoch) {
+	cc.waitForEpochOffset(ctx, e, 0)
+}
+
+// WaitForEpochPropDelay returns propDelay time after the start of the epoch, or when ctx is done.
+func (cc *chainClock) WaitForEpochPropDelay(ctx context.Context, e abi.ChainEpoch) {
+	cc.waitForEpochOffset(ctx, e, cc.propDelay)
+}
+
+// waitNextEpochOffset returns when time is offset past the start of the epoch, or ctx is done.
+func (cc *chainClock) waitForEpochOffset(ctx context.Context, e abi.ChainEpoch, offset time.Duration) {
+	epochStart := cc.StartTimeOfEpoch(e).Add(offset)
```

Comment:
nit: the name is not exactly right int the case there is an offset.  Hard to think of a better one though.